cmd/upgrade/app/v1alpha1/result.go

```diff
@@ -0,0 +1,278 @@
+	if len(urb.Tasks) == 0 {
+		validationErrs = append(validationErrs,
+			errors.New("missing tasks"))
+	}
+	if len(validationErrs) != 0 {
+		urb.errors = append(urb.errors, validationErrs...)
+		return errors.Errorf("validation error(s) found: %v", validationErrs)
+	}
+	return nil
+}
+
+// GetOrCreate builds a new instance of UpgradeResult with the
+// helps of UpgradeResultGetOrCreateBuilder. Upgrade result cr
+// is required to maintain resiliency in upgrade.
+func (urb *UpgradeResultGetOrCreateBuilder) GetOrCreate() (
+	res *upgrade.UpgradeResult, err error) {
+	err = urb.validate()
+	if err != nil {
+		return nil,
+			errors.Wrapf(err, "failed to get or create upgrade result: %s", urb)
+	}
+	l := labelJobName + "=" + urb.SelfName +
+		"," + labelItemName + "=" + urb.ResourceDetails.Name +
+		"," + labelItemNamespace + "=" + urb.ResourceDetails.Namespace +
+		"," + labelItemKind + "=" + urb.ResourceDetails.Kind
+	opts := metav1.ListOptions{
+		LabelSelector: l,
+	}
+	urList, err := upgraderesult.KubeClient(
+		upgraderesult.WithNamespace(urb.SelfNamespace)).
+		List(opts)
+	if err != nil {
+		return nil,
+			errors.Wrapf(err, "failed to get or create upgrade result: %s", urb)
+	}
+	switch urCount := len(urList.Items); urCount {
+	case 0:
+		ur, err := urb.buildUpgradeResult()
+		if err != nil {
+			return nil,
+				errors.Wrapf(err, "failed to get or create upgrade result: %s", urb)
+		}
+
+		urb.UpgradeResult.object, err = upgraderesult.KubeClient(
+			upgraderesult.WithNamespace(urb.SelfNamespace)).
+			Create(ur)
+		if err != nil {
+			return nil,
+				errors.Wrapf(err, "failed to get or create upgrade result: failed to create: %s", urb)
+		}
+		return urb.UpgradeResult.object, nil
+	case 1:
+		return &urList.Items[0], nil
+	default:
+		return nil,
+			errors.Errorf(
+				"failed to get or create upgrade result builder: more than one upgrade result instances were found for resource {%v}: upgrade result instances {%v}",
+				urb.ResourceDetails, urList)
+	}
+}
+
+// buildUpgradeResult returns UpgradeResult Object
+func (urb *UpgradeResultGetOrCreateBuilder) buildUpgradeResult() (
+	*upgrade.UpgradeResult, error) {
+	tm, err := urb.getTypeMeta()
+	if err != nil {
+		return nil, err
+	}
+
+	om, err := urb.getObjectMeta()
+	if err != nil {
+		return nil, err
+	}
+
+	return upgraderesult.NewBuilder().
+		WithTypeMeta(*tm).
+		WithObjectMeta(*om).
+		WithResultConfig(*urb.ResourceDetails, urb.UpgradeConfig.Data...).
+		WithTasks(urb.Tasks...).
+		Build()
+}
+
+// getTypeMeta returns metav1.TypeMeta for upgrade result cr
+func (urb *UpgradeResultGetOrCreateBuilder) getTypeMeta() (
+	tm *metav1.TypeMeta, err error) {
+	return typemeta.NewBuilder().
+		WithKind("UpgradeResult").
+		WithAPIVersion("openebs.io/v1alpha1").
+		Build()
+}
+
+// getObjectMeta returns metav1.ObjectMeta for upgrade result cr.
+func (urb *UpgradeResultGetOrCreateBuilder) getObjectMeta() (
+	tm *metav1.ObjectMeta, err error) {
+	name := urb.SelfName + "-" + rand.String(4)
```

Comment:
can we make use of `generateName` property of object meta?

Reply:
done